### PR TITLE
feat: add atomic_write_file, atomic_write_file_with_retries, and get_atomic_path

### PR DESCRIPTION
**Pull request by dsherret**
Moved from the CLI and update to use https://github.com/dsherret/sys_traits

### CODE REVIEW

File: src/fs.rs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+) -> std::io::Result<()> {
+  fn atomic_write_file_raw<TSys: FsOpen + FsRename + FsRemoveFile>(
+    sys: &TSys,
+    temp_file_path: &Path,
+    file_path: &Path,
+    data: &[u8],
+    mode: u32,
+  ) -> std::io::Result<()> {
+    let mut file = sys.fs_open(temp_file_path, &OpenOptions::write())?;
+    file.fs_file_set_permissions(mode)?;
```
Comment (dsherret) [inline src/fs.rs:112]: I copied this from the CLI, but I kind of feel like we should only be setting the file permissions on open (the OpenOptions allow for that). Thoughts?

File: src/fs.rs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+}
+
+pub fn atomic_write_file_with_retries<
+  TSys: FsCreateDirAll
+    + FsMetadata
+    + FsOpen
+    + FsRemoveFile
+    + FsRename
+    + ThreadSleep
+    + SystemRandom,
```
Comment (dsherret) [inline src/fs.rs:66]: I know this seems slightly out of control, but it makes using this crate much easier. A main problem atm is Rust doesn't have a way to create aliases of traits, so there's no way to just say "`deno_path_util` uses these traits under the name `DenoPathUtilSys`, so now accept `TSys: DenoPathUtilSys`". Yes, we could do `pub trait DenoPathUtilSys: FsCreateDirAll + ... {}`, but a huge negative to that is then people need to implement the `DenoPathUtilSys` crate for file system implementation and that might not always be possible if the file system implementation lives in a different crate... so we add some verbosity here for a better experience for consumers.

File: src/lib.rs [hunk 3]
```diff
@@ -5,13 +5,15 @@
...
 
-use std::io::ErrorKind;
 use std::path::Component;
 use std::path::Path;
 use std::path::PathBuf;
+use sys_traits::SystemRandom;
 use thiserror::Error;
 use url::Url;
 
+pub mod fs;
```
Comment (dsherret) [inline src/lib.rs:15]: This should arguably be in a separate crate, but I don't think it's worth the additional maintenance burden.

File: src/fs.rs (hunk 1 above)
Comment (bartlomieju) [reply src/fs.rs:112]: I'm not sure I understand what you mean by "setting the file permissions on open". It seems desirable that the file is written with proper permissions.

File: src/lib.rs (hunk 3 above)
Comment (bartlomieju) [reply src/lib.rs:15]: I agree it's not worth it.

**Review by bartlomieju — APPROVED** [review]

File: src/fs.rs (hunk 1 above)
Comment (dsherret) [reply src/fs.rs]: @bartlomieju I did a comimt to show it.

**Pull request merged**